kubectl alpha debug fails with "panic: runtime error: index out of range"

**What happened**:
When using `kubectl alpha debug` to create a debug container for a pod that has multiple init containers or ephemeral containers, `kubectl alpha debug` fails with `panic: runtime error: index out of range`.
```
# kubectl alpha debug myapp-pod --image=busybox
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.containerNameToRef(0xc00010a400, 0x18)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:547 +0x299
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).computeDebugContainerName(0xc0000d8fc0, 0xc00010a400, 0x20, 0x19aa0a0)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:528 +0x4b
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).generateDebugContainer(0xc0000d8fc0, 0xc00010a400, 0x2)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:389 +0x39
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).debugByEphemeralContainer(0xc0000d8fc0, 0x1e93060, 0xc0000be010, 0xc00010a400, 0x1ed9a37800000000, 0xc000a079b8, 0xc00062a000, 0xc000650718, 0x4)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:359 +0x386
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).visitPod(0xc0000d8fc0, 0x1e93060, 0xc0000be010, 0xc00010a400, 0x0, 0x0, 0x0, 0x1000100, 0x0)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:343 +0xc5
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).Run.func1(0xc0007060e0, 0x0, 0x0, 0x0, 0x0)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:284 +0xd7
```

**What you expected to happen**:
`kubectl alpha debug` should create a debug container in pod `myapp-pod` successfully.

**How to reproduce it (as minimally and precisely as possible)**:
The EphemeralContainers feature gate must be enabled by using `--feature-gates=EphemeralContainers=true` flag in your cluster.
```
# cat myapp-pod.yaml
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
spec:
  containers:
  - name: myapp-container
    image: busybox
    command: ['sh', '-c', 'echo The app is running! && sleep 1d']
  initContainers:
  - name: init-container-1
    image: busybox
    command: ['sh', '-c', 'sleep 1d']
  - name: init-container-2
    image: busybox
    command: ['sh', '-c', 'sleep 1d']

# kubectl apply -f myapp-pod.yaml
# kubectl alpha debug myapp-pod --image=busybox
panic: runtime error: index out of range [1] with length 1
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
# kubectl version 
Client Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.0-alpha.0.996+74b04768b6d1a0", GitCommit:"74b04768b6d1a087c106828f45aeed2f84fa971e", GitTreeState:"clean", BuildDate:"2020-09-06T15:37:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.0-alpha.0.996+74b04768b6d1a0", GitCommit:"74b04768b6d1a087c106828f45aeed2f84fa971e", GitTreeState:"clean", BuildDate:"2020-09-06T15:37:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
NAME="CentOS Linux"
VERSION="7 (Core)"
- Kernel (e.g. `uname -a`):
Linux node-1 3.10.0-957.el7.x86_64 #1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig cli

--- a/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
@@ -485,6 +485,8 @@ func (o *DebugOptions) generatePodCopyWithDebugContainer(pod *corev1.Pod) (*core
 		},
 		Spec: *pod.Spec.DeepCopy(),
 	}
+	// set EphemeralContainers to nil so that the copy of pod can be created
+	copied.Spec.EphemeralContainers = nil
 	// change ShareProcessNamespace configuration only when commanded explicitly
 	if o.shareProcessedChanged {
 		copied.Spec.ShareProcessNamespace = pointer.BoolPtr(o.ShareProcesses)
@@ -544,11 +546,11 @@ func containerNameToRef(pod *corev1.Pod) map[string]*corev1.Container {
 		names[ref.Name] = ref
 	}
 	for i := range pod.Spec.InitContainers {
-		ref := &pod.Spec.Containers[i]
+		ref := &pod.Spec.InitContainers[i]
 		names[ref.Name] = ref
 	}
 	for i := range pod.Spec.EphemeralContainers {
-		ref := &pod.Spec.Containers[i]
+		ref := (*corev1.Container)(&pod.Spec.EphemeralContainers[i].EphemeralContainerCommon)
 		names[ref.Name] = ref
 	}
 	return names

--- a/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug_test.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/debug/debug_test.go
@@ -153,6 +153,74 @@ func TestGenerateDebugContainer(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "pod with init containers",
+			opts: &DebugOptions{
+				Image:      "busybox",
+				PullPolicy: corev1.PullIfNotPresent,
+			},
+			pod: &corev1.Pod{
+				Spec: corev1.PodSpec{
+					InitContainers: []corev1.Container{
+						{
+							Name: "init-container-1",
+						},
+						{
+							Name: "init-container-2",
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name: "debugger",
+						},
+					},
+				},
+			},
+			expected: &corev1.EphemeralContainer{
+				EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+					Name:                     "debugger-1",
+					Image:                    "busybox",
+					ImagePullPolicy:          "IfNotPresent",
+					TerminationMessagePolicy: "File",
+				},
+			},
+		},
+		{
+			name: "pod with ephemeral containers",
+			opts: &DebugOptions{
+				Image:      "busybox",
+				PullPolicy: corev1.PullIfNotPresent,
+			},
+			pod: &corev1.Pod{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name: "debugger",
+						},
+					},
+					EphemeralContainers: []corev1.EphemeralContainer{
+						{
+							EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+								Name: "ephemeral-container-1",
+							},
+						},
+						{
+							EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+								Name: "ephemeral-container-2",
+							},
+						},
+					},
+				},
+			},
+			expected: &corev1.EphemeralContainer{
+				EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+					Name:                     "debugger-1",
+					Image:                    "busybox",
+					ImagePullPolicy:          "IfNotPresent",
+					TerminationMessagePolicy: "File",
+				},
+			},
+		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
 			tc.opts.IOStreams = genericclioptions.NewTestIOStreamsDiscard()
@@ -567,6 +635,110 @@ func TestGeneratePodCopyWithDebugContainer(t *testing.T) {
 					Containers: []corev1.Container{
 						{
 							Name: "debugger-1",
+						},
+					},
+				},
+			},
+			expected: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "debugger",
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name: "debugger-1",
+						},
+						{
+							Name:                     "debugger-2",
+							Image:                    "busybox",
+							ImagePullPolicy:          corev1.PullIfNotPresent,
+							TerminationMessagePolicy: corev1.TerminationMessageReadFile,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "pod with init containers",
+			opts: &DebugOptions{
+				CopyTo:     "debugger",
+				Image:      "busybox",
+				PullPolicy: corev1.PullIfNotPresent,
+			},
+			pod: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "target",
+				},
+				Spec: corev1.PodSpec{
+					InitContainers: []corev1.Container{
+						{
+							Name: "init-container-1",
+						},
+						{
+							Name: "init-container-2",
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name: "debugger-1",
+						},
+					},
+				},
+			},
+			expected: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "debugger",
+				},
+				Spec: corev1.PodSpec{
+					InitContainers: []corev1.Container{
+						{
+							Name: "init-container-1",
+						},
+						{
+							Name: "init-container-2",
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name: "debugger-1",
+						},
+						{
+							Name:                     "debugger-2",
+							Image:                    "busybox",
+							ImagePullPolicy:          corev1.PullIfNotPresent,
+							TerminationMessagePolicy: corev1.TerminationMessageReadFile,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "pod with ephemeral containers",
+			opts: &DebugOptions{
+				CopyTo:     "debugger",
+				Image:      "busybox",
+				PullPolicy: corev1.PullIfNotPresent,
+			},
+			pod: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "target",
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name: "debugger-1",
+						},
+					},
+					EphemeralContainers: []corev1.EphemeralContainer{
+						{
+							EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+								Name: "ephemeral-container-1",
+							},
+						},
+						{
+							EphemeralContainerCommon: corev1.EphemeralContainerCommon{
+								Name: "ephemeral-container-2",
+							},
 						},
 					},
 				},

